docs: document the process of running the tests in the cloud machines (#2404)
elastic/e2e-testing: e2e/TROUBLESHOOTING.md

Each test suite's documentation should contain the specifics to run the tests, but it's summarises to executing `go test` or `godog` in the right directory.

### Running the tests on the Cloud machines
On CI, we are running the Elastic Stack and all test suites in AWS instances, so whenever a build failed we would need to access those machines and inspect the state of the machine: logs, files, containers... For that, we are enabling SSH access to those ephemeral machines, which will be kept for debugging purpose if and only if the `DEVELOPER_MODE` environment variable is set at the Jenkinsfile. In the UI of Jenkins, you can enable it using the `DEVELOPER_MODE` input argument, checking it to true (default is false). After the build finishes, the cloud instances won't be destroyed.

> But you must first understand that there are two types of Cloud machines: 1) the VM running the stack, and 2) the VMs where the Elastic Agent will be installed and enrolled into the stack.

Reviewer: [Suggestion]
It's in part personal opinion, but I find it easier to read when the items are each one on a separated line:

```suggestion
But you must first understand that there are two types of Cloud machines: 
1) the VM running the stack, and 
2) the VMs where the Elastic Agent will be installed and enrolled into the stack.
```

To get the IP address of the VMs, please go to the Jenkins'm BlueOcean UI of the job you manually triggered with `DEVELOPER_MODE=true`, and look up the **Deploy Test Infra** stage. Under its steps, you will see different Ansible executions to provision the VM. Please open any of them and look for any log entry containing an IP address:

```shell
[2022-04-21T05:20:27.133Z] TASK [Gathering Facts] *********************************************************
[2022-04-21T05:20:29.541Z] ok: [3.144.74.102]
```

The IP address of that VM is `3.144.74.102`.

For the agent VMs it's exactly the same, but looking up the **parallel stages** right after the Test Infra. Again, look for any Ansible task containing an IP address, as shown above.

> Once you have both IP addresses, one for the stack and one for the agent in the OS/Arch you are interested in, please open two terminals, one for each. Then SSH into the machines using: 1) the public SSH key you have in your Github account, and 2) the right user for the machine, as described [here](https://github.com/elastic/e2e-testing/blob/4517dfa134844f720139d6bab3955cc8d9c6685c/.ci/.e2e-platforms.yaml#L2-L42).

Reviewer: [Suggestion]
It's in part personal opinion, but I find it easier to read when the items are each one on a separated line:

```suggestion
Once you have both IP addresses, one for the stack and one for the agent in the OS/Arch you are interested in, please open two terminals, one for each. Then SSH into the machines using: 
1) the public SSH key you have in your Github account, and 
2) the right user for the machine, as described [here](https://github.com/elastic/e2e-testing/blob/4517dfa134844f720139d6bab3955cc8d9c6685c/.ci/.e2e-platforms.yaml#L2-L42).
```

An example of how to SSH in the machine, having multiple SSH keys under the `.ssh` directory, and connecting with the `admin` user because it's a Debian machine:

```shell
ssh -i ~/.ssh/id_rsa_elastic.pub -vvvv admin@18.188.242.30
```

#### Running the tests for the Elastic Agent
Once you have SSH'ed into the right agent VM (ex. CentOS 8 ARM), and checked that the stack is in a valid state (the API token of the Fleet Server has not expired, otherwise you must recreate the entire stack), you can run the tests with a simple command, but you must first set the environment so that the agent is able to connect to the remote stack, which lives in another VM. Just source the `.env` file that the test framework created:

```shell
# log in as root
sudo su -
# move to the project directory
cd /home/${USER}/e2e-testing
# load the environment
source .env
# verify variables
env
```

The env should contain those variables needed for enroling the agent, such as `ELASTICSEARCH_URL` and `FLEET_SERVER_URL`, among others.

> Now you can run the tests, specifying the tags you are interested. Please use the tags in the feature files, where the test framework defines one at the top level, for running an entire feature file, and per scenario, so that it's possible to tell the test runner to run one or multiple scenarios. More about Cucumber tags in [here](https://github.com/cucumber/godog#tags).
> 

Reviewer: [Suggestion]
I'd suggest to explain better what are the tags and how to find them. [The docs ](https://github.com/cucumber/godog#tags) explain only how to use them, not how to define them. I had a hard time to find where the tags were. So this is my suggestion:

```suggestion
Now you can run the tests, specifying the tags you are interested. Please use the tags in the feature files.
The test framework defines a tag at the top level, for running an entire feature file, and one per scenario, so that it's possible to tell the test runner to run one or multiple scenarios.
The tags are the `@tag-name` [found at the very top of the `.feature` files](https://github.com/elastic/e2e-testing/blob/main/e2e/_suites/fleet/features/fleet_mode_agent.feature#L1) or [above each individual test](https://github.com/elastic/e2e-testing/blob/main/e2e/_suites/fleet/features/system_integration.feature#L11).
More about Cucumber tags in [here](https://github.com/cucumber/godog#tags).

```

```shell
# log in as root
sudo su -
# move to the project directory
cd /home/${USER}/e2e-testing
TAGS="deploy-system_integration-with-diskio" TIMEOUT_FACTOR=5 LOG_LEVEL=TRACE PROVIDER=remote make -C e2e/_suites/fleet functional-test
```

- TAGS: it uses a tag from the `system_integration.feature` file, because we want to run just one scenario.
- LOG_LEVEL: keep it as TRACE to see everything.
- TIMEOUT_FACTOR: this number will multiply the default timeouts when waiting for states, such as waiting for an agent to be online. Default is 3 minutes.

> - PROVIDER: use `REMOTE` so that it will connect to the remote stack you already provisioned.

Author: I think this suggestion has been broken by Github's UI, as it affects a "```" block

> - PROVIDER: use `REMOTE` so that it will connect to the remote stack you already provisioned.

Reviewer: yes, it was

> - PROVIDER: use `REMOTE` so that it will connect to the remote stack you already provisioned.

Author: @AndersonQ can we merge this? You can add your suggestion as a follow-up PR

You may need to SSH into this machine to recreate the stack in the case the API Token used for Fleet Server is [expired after 1 hour](https://www.elastic.co/guide/en/elasticsearch/reference/current/security-settings.html#token-service-settings), 

If you want to recreate the stack, please run the following command, which will run the tests but without any valid scenario (see `TAGS` variable):

```shell
# login as root
sudo su -
# stop previous stack
docker-compose -f ~/.op/compose/profiles/fleet/docker-compose.yml down --remove-orphans
# bootstrap the new stack
TAGS="non-existing-tag" TIMEOUT_FACTOR=3 LOG_LEVEL=TRACE DEVELOPER_MODE=true make -C e2e/_suites/fleet functional-test
```

#### The Agent VMs
There different VM flavours that you can use to run the Elastic Agent and enrol it into the Stack: Debian, CentOS, SLES15, Oracle Linux... using AMD and ARM as architecture. You can find the full reference of the platform support [here](https://github.com/elastic/e2e-testing/blob/4517dfa134844f720139d6bab3955cc8d9c6685c/.ci/.e2e-platforms.yaml#L2-L42).

> In these VMs, the test framework will download a binary to install the Elastic Agent (TAR files, DEB/RPM packages...), and will execute the different agent commands to install, enrol, uninstall, etc.

Reviewer 2: ```suggestion
In these VMs, the test framework will download a binary to install the Elastic Agent (TAR files, DEB/RPM packages...), and will execute the different agent commands to install, enroll, uninstall, etc.
```

#### Getting SSH access to the VMs

> To access the machines, you must be allowed to do so first, and for that, please submit a PR adding your Github username in alphabetical order to [this file](../.ci/ansible/github-ssh-keys), keeping a blank line as file ending. The user to access each EC2 used on the tests can be found [here](https://github.com/elastic/e2e-testing/blob/main/.ci/.e2e-platforms.yaml). When submitting the pul request with your user, please remember to add the right backport labels (ex. `backport-v8.2.0`) so that you will be able to SSH into the supported maintenance branches.

Reviewer 2: ```suggestion
To access the machines, you must be allowed to do so first, and for that, please submit a PR adding your Github username in alphabetical order to [this file](../.ci/ansible/github-ssh-keys), keeping a blank line as file ending. The user to access each EC2 used on the tests can be found [here](https://github.com/elastic/e2e-testing/blob/main/.ci/.e2e-platforms.yaml). When submitting the pull request with your user, please remember to add the right backport labels (ex. `backport-v8.2.0`) so that you will be able to SSH into the supported maintenance branches.
```

> To get the IP address of the VMs, please go to the Jenkins'm BlueOcean UI of the job you manually triggered with `DEVELOPER_MODE=true`, and look up the **Deploy Test Infra** stage. Under its steps, you will see different Ansible executions to provision the VM. Please open any of them and look for any log entry containing an IP address:

Reviewer 2: ```suggestion
To get the IP address of the VMs, please go to the Jenkins BlueOcean UI of the job you manually triggered with `DEVELOPER_MODE=true`, and look up the **Deploy Test Infra** stage. Under its steps, you will see different Ansible executions to provision the VM. Please open any of them and look for any log entry containing an IP address:
```

1) the public SSH key you have in your Github account, and 
2) the right user for the machine, as described [here](https://github.com/elastic/e2e-testing/blob/4517dfa134844f720139d6bab3955cc8d9c6685c/.ci/.e2e-platforms.yaml#L2-L42).

An example of how to SSH in the machine, having multiple SSH keys under the `.ssh` directory, and connecting with the `admin` user because it's a Debian machine:

```shell
ssh -i ~/.ssh/id_rsa_elastic.pub -vvvv admin@18.188.242.30
```

#### Running the tests for the Elastic Agent
Once you have SSH'ed into the right agent VM (ex. CentOS 8 ARM), and checked that the stack is in a valid state (the API token of the Fleet Server has not expired, otherwise you must recreate the entire stack), you can run the tests with a simple command, but you must first set the environment so that the agent is able to connect to the remote stack, which lives in another VM. Just source the `.env` file that the test framework created:

```shell
# log in as root
sudo su -
# move to the project directory
cd /home/${USER}/e2e-testing
# load the environment
source .env
# verify variables
env
```

> The env should contain those variables needed for enroling the agent, such as `ELASTICSEARCH_URL` and `FLEET_SERVER_URL`, among others.

Reviewer 2: ```suggestion
The env should contain those variables needed for enrolling the agent, such as `ELASTICSEARCH_URL` and `FLEET_SERVER_URL`, among others.
```
